openedx_learning/core/components/models.py

```diff
@@ -0,0 +1,403 @@
+    act of publishing means we're marking a ContentVersion as the official,
+    ready-for-use version of this Component.
+    """
+
+    publish_log_entry = models.ForeignKey(PublishLogEntry, on_delete=models.CASCADE)
+    component = models.ForeignKey(Component, on_delete=models.RESTRICT)
+    component_version = models.ForeignKey(
+        ComponentVersion, on_delete=models.RESTRICT, null=True
+    )
+
+
+class PublishedComponent(models.Model):
+    """
+    For any given Component, what is the currently published ComponentVersion.
+
+    It may be possible for a Component to exist only as a Draft (and thus not
+    show up in this table). There is only ever one published ComponentVersion
+    per Component at any given time.
+
+    TODO: Do we need to create a (redundant) title field in this model so that
+    we can more efficiently search across titles within a LearningPackage?
+    Probably not an immediate concern because the number of rows currently
+    shouldn't be > 10,000 in the more extreme cases.
+    """
+
+    component = models.OneToOneField(
+        Component, on_delete=models.RESTRICT, primary_key=True
+    )
+    component_version = models.OneToOneField(
+        ComponentVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+    )
+    component_publish_log_entry = models.ForeignKey(
+        ComponentPublishLogEntry,
+        on_delete=models.RESTRICT,
+    )
+
+    class Meta:
+        verbose_name = "Published Component"
+        verbose_name_plural = "Published Components"
+
+
+class Content(models.Model):
+    """
+    This is the most basic piece of raw content data, with no version metadata.
+
+    Content stores data in an immutable Binary BLOB `data` field. This data is
+    not auto-normalized in any way, meaning that pieces of content that are
+    semantically equivalent (e.g. differently spaced/sorted JSON) will result in
+    new entries. This model is intentionally ignorant of what these things mean,
+    because it expects supplemental data models to build on top of it.
+
+    Two Content instances _can_ have the same hash_digest if they are of
+    different MIME types. For instance, an empty text file and an empty SRT file
+    will both hash the same way, but be considered different entities.
+
+    The other fields on Content are for data that is intrinsic to the file data
+    itself (e.g. the size). Any smart parsing of the contents into more
+    structured metadata should happen in other models that hang off of Content.
+
+    Content models are not versioned in any way. The concept of versioning only
+    exists at a higher level.
+
+    Since this model uses a BinaryField to hold its data, we have to be careful
+    about scalability issues. For instance, video files should not be stored
+    here directly. There is a 10 MB limit set for the moment, to accomodate
+    things like PDF files and images, but the itention is for the vast majority
+    of rows to be much smaller than that.
```

Comment:
How _should_ data files over 10 MB be stored?

I would almost prefer a much lower limit like 250 kB and an integrated way to store and reference larger Content instances as objects on S3. I suspect you may already have implemented or planned something that but I forgot or don't see it yet.

Reply:
You're right, I've punted on this issue. I came to the same conclusion that you did: S3 objects (probably being django-storages) + database references, probably through an optional FileField on Content.

Captured this in https://github.com/openedx/openedx-learning/issues/29